Sync with development branch

alvarolopez opened the pull request.

alvarolopez committed:
Initial spot API extension implementation

This is an extension for the OpenStack API V2.1 that makes possible to
mark an instance being a spot.

(cherry picked from commit e20814b8da7a1f501b20ef3af8417cb7472b1736)

This commit has been forklifted from the nova repository

alvarolopez committed:
Move modules to our namespace

The api code was forklifted from the old nova repository, and is not
adapted to our namespace. Lets move the files around, and create the
appropriate packages.

alvarolopez committed:
Adapt code to the Liberty release

Starting from liberty, the API extension loading has changed and now is
based on Stevedore and entry points. Adapt the code to work with this
new setup.

alvarolopez committed:
Create a HostStatePartial and a new HostManager classes

In order to track the resources consumed by the preemptible instances we
introduce the HostStatePartial class inheriting from the original
HostStace class. It is able to return two different host states:

 - A partial state, taking into account only normal instances (i.e.
   considering the resources consumed by the preemptible instances as
   free).
 - A full state, taking into account _all_ instances inside a host.

The new HostManager class extends the original manager (that is
nova.scheduler.host_manager.HostManager), adding a new method for
obtaining the host partial states.

alvarolopez committed:
fix W291 error from cookiecutter

alvarolopez committed:
api: raise HTTP NotImplemented instead of returning empty

In the index method of the preemptible instances controller we were
returning an empty list, but semantically this was not correct. Raise
the proper error here, so that the client knows that we do not implement
this for now.

alvarolopez committed:
Update requirements and test-requirements

Add missing dependencies, and align with upstream nova stable/liberty
versions.

alvarolopez committed:
Update tox.ini with upstream changes

Modify tox configuration file so that it is more aligned with the nova
upstream stable/liberty version.

alvarolopez committed:
Add configuration for oslo_config_generator

Add a file containing the configuration for the generator, as long as
a module holding all the options.

alvarolopez committed:
tests: fix coverage tests

When aligning tox.ini with the upstream version, the coverage was
running against the "nova" package, not opie. Also, the .coveragerc file
contained a sintax error.

alvarolopez commented:
@orviz the unittests are failing, as a requirement is not installed: nova needs to be installed as specified in https://github.com/indigo-dc/opie/pull/3/commits/f1156ddf2692974b9af0d80c1bdfc97c08947c1f#diff-b91f3d5bd63fcd17221b267e851608e8R18. We cannot add it in the `requirements.txt` file as it is not in PyPi.

Regarding the codestyle, there is an error regarding the locale that I guess is unrelated with opie.

alvarolopez committed:
Introduce a preemptible instances filter scheduler

This change introduces a new FilterScheduler, starting from the original
nova.scheduler.filter_scheduler.FilterScheduler. This new class is able
to schedule normal instances by preempting one or several preemptible
instances if there are not enough resources available for an incoming
request.

In order to do the scheduling of the preemptible and normal instances,
the new FilterScheduler will perform the normal filter phase taking into
account these two states:

 - normal state, when scheduling preemptible instances.
 - partial state, when scheduling a normal instance.

This way, it will be possible to schedule a normal instance even if
there are no resources available (consumed by preemptible instances) in
the filtering phase.

When the filtering phase has finished, the weighing phase will start.
This phase will take into account normal state of the hosts, regardless
of the request being a preemptible or a normal one. This way, it is
possible to implement several weighers that will ensure that the best
node is selected, so that:

 - It will be possible to give priority to hosts without preemptible
   instances (so that no instance needs to be killed since this
   operation will incur in a penalty for the provider)

 - It will be possible to select the node with the preemptible that will
   cost less to be killed.

 - It will be possible to implement more advanced weighers, that will
   take into account priorities, etc. This is however out of the scope
   of this BP.

Once the host is selected via the modified filtering and weighing phase,
the scheduler will check if the host resources are overcommited and if
there are preemptible instances to be terminated. If so, the scheduler
will initiate a new weighing phase, so as to select which preemptible
instances inside that host need to be terminated.

alvarolopez committed:
Py3: do not run all tests in Python 3

Not all nova tests are able to run in Python 3, therefore we need to
specify them one by one. Currently, no tests are able to run in Py3.

orviz commented:
@alvarolopez unittest check is fine after adding the nova requirement to jenkins build job

codestyle is still failing since it does not manage to get upper-constraints.txt

orviz commented:
@alvarolopez all fine using tox command directly

alvarolopez committed:
api: add unit testing

This change adds unit testing for the API extensions, updating the
required modules for the tests.

alvarolopez committed:
tests: remove tests.base module

We are going to rely on the nova tests for now, so we can safely remove
the almost empty and useless tests.base module for now.
